fix google benchmark deprecations

When building with a recent Google bench and `MDSPAN_ENABLE_BENCHMARKS=ON`:

```
benchmarks/sum/sum_3d_left.cpp:52:16: error: 'DoNotOptimize<int *>' is deprecated: The const-ref version of this method can permit undesired compiler optimizations in benchmarks [-Werror,-Wdeprecated-declarations]
[build]    52 |     benchmark::DoNotOptimize(s.data_handle());
```

This was deprecated by google benchmark due to its inability to guarantee certain optimizations. The solution is to use temporaries as in https://github.com/google/benchmark/pull/1568/files

--- a/benchmarks/copy/copy_layout_stride.cpp
+++ b/benchmarks/copy/copy_layout_stride.cpp
@@ -41,8 +41,10 @@ void BM_MDSpan_Copy_2D_right(benchmark::State& state, MDSpan, DynSizes... dyn) {
           dest(i, j) = s(i, j);
       }
     }
-    benchmark::DoNotOptimize(s.data_handle());
-    benchmark::DoNotOptimize(dest.data_handle());
+    auto sdata = s.data_handle();
+    auto ddata = dest.data_handle();
+    benchmark::DoNotOptimize(sdata);
+    benchmark::DoNotOptimize(ddata);
   }
   state.SetBytesProcessed(s.size() * sizeof(value_type) * state.iterations());
 }
@@ -78,8 +80,10 @@ void BM_MDSpan_Copy_2D_stride(benchmark::State& state, MDSpan, LayoutMapping map
         dest(i, j) = s(i, j);
       }
     }
-    benchmark::DoNotOptimize(s.data_handle());
-    benchmark::DoNotOptimize(dest.data_handle());
+    auto sdata = s.data_handle();
+    auto ddata = dest.data_handle();
+    benchmark::DoNotOptimize(sdata);
+    benchmark::DoNotOptimize(ddata);
   }
   state.SetBytesProcessed(s.size() * sizeof(value_type) * state.iterations());
 }
@@ -146,8 +150,10 @@ void BM_MDSpan_Copy_2D_stride_diff_map(benchmark::State& state,
         dest(i, j) = src(i, j);
       }
     }
-    benchmark::DoNotOptimize(src.data_handle());
-    benchmark::DoNotOptimize(dest.data_handle());
+    auto sdata = src.data_handle();
+    auto ddata = dest.data_handle();
+    benchmark::DoNotOptimize(sdata);
+    benchmark::DoNotOptimize(ddata);
   }
   state.SetBytesProcessed(src.extent(0) * src.extent(1) * sizeof(value_type) * state.iterations());
 }

--- a/benchmarks/sum/sum_3d_left.cpp
+++ b/benchmarks/sum/sum_3d_left.cpp
@@ -49,7 +49,8 @@ void BM_MDSpan_Sum_3D_left(benchmark::State& state, MDSpan, DynSizes... dyn) {
       }
     }
     benchmark::DoNotOptimize(sum);
-    benchmark::DoNotOptimize(s.data_handle());
+    auto sdata = s.data_handle();
+    benchmark::DoNotOptimize(sdata);
   }
   state.SetBytesProcessed(s.size() * sizeof(value_type) * state.iterations());
 }

--- a/benchmarks/sum/sum_3d_right.cpp
+++ b/benchmarks/sum/sum_3d_right.cpp
@@ -45,7 +45,8 @@ void BM_MDSpan_Sum_3D_right(benchmark::State& state, MDSpan, DynSizes... dyn) {
 
   for (auto _ : state) {
     benchmark::DoNotOptimize(s);
-    benchmark::DoNotOptimize(s.data_handle());
+    auto sdata = s.data_handle();
+    benchmark::DoNotOptimize(sdata);
     value_type sum = 0;
     for(index_type i = 0; i < s.extent(0); ++i) {
       for (index_type j = 0; j < s.extent(1); ++j) {

--- a/benchmarks/sum/sum_submdspan_right.cpp
+++ b/benchmarks/sum/sum_submdspan_right.cpp
@@ -45,7 +45,8 @@ void BM_MDSpan_Sum_Subspan_3D_right(benchmark::State& state, MDSpan, DynSizes...
 
   for (auto _ : state) {
     benchmark::DoNotOptimize(s);
-    benchmark::DoNotOptimize(s.data_handle());
+    auto sdata = s.data_handle();
+    benchmark::DoNotOptimize(sdata);
     value_type sum = 0;
     using index_type = typename MDSpan::index_type;
     for(index_type i = 0; i < s.extent(0); ++i) {
@@ -58,7 +59,7 @@ void BM_MDSpan_Sum_Subspan_3D_right(benchmark::State& state, MDSpan, DynSizes...
       }
     }
     benchmark::DoNotOptimize(sum);
-    benchmark::DoNotOptimize(s.data_handle());
+    benchmark::DoNotOptimize(sdata);
   }
   state.SetBytesProcessed(s.size() * sizeof(value_type) * state.iterations());
 }
@@ -154,7 +155,8 @@ void BM_MDSpan_Sum_Subspan_MD_right(benchmark::State& state, MDSpan, DynSizes...
     value_type sum = 0;
     _impl::_do_sum_submdspan(sum, s);
     benchmark::DoNotOptimize(sum);
-    benchmark::DoNotOptimize(s.data_handle());
+    auto sdata = s.data_handle();
+    benchmark::DoNotOptimize(sdata);
   }
   state.SetBytesProcessed(s.size() * sizeof(value_type) * state.iterations());
 }

--- a/benchmarks/tiny_matrix_add/tiny_matrix_add.cpp
+++ b/benchmarks/tiny_matrix_add/tiny_matrix_add.cpp
@@ -47,9 +47,11 @@ void BM_MDSpan_TinyMatrixSum_right(benchmark::State& state, MDSpan, DynSizes... 
 
   for (auto _ : state) {
     benchmark::DoNotOptimize(o);
-    benchmark::DoNotOptimize(o.data_handle());
+    auto odata = o.data_handle();
+    benchmark::DoNotOptimize(odata);
     benchmark::DoNotOptimize(s);
-    benchmark::DoNotOptimize(s.data_handle());
+    auto sdata = s.data_handle();
+    benchmark::DoNotOptimize(sdata);
     for(index_type i = 0; i < s.extent(0); i ++) {
       for(index_type j = 0; j < s.extent(1); j ++) {
         for(index_type k = 0; k < s.extent(2); k ++) {

